packages/scratch-vm/src/extensions/scratch3_spotify/index.ts

```diff
@@ -0,0 +1,696 @@
+                    return reject();
+                }
+
+                if (res.statusCode !== 200) {
+                    console.error("Spotify token error: " + res.statusCode);
+                    return reject();
+                }
+
+                // success
+                let token: AccessToken = {
+                    expirationTime: currentTimeSec() + 3600,
+                    value: JSON.parse(body).token,
+                };
+                resolve(token);
+            }
+        );
+    });
+};
+
+const refreshAccessTokenIfNeeded = (
+    token: AccessToken
+): Promise<AccessToken> => {
+    return new Promise((resolve, reject) => {
+        if (currentTimeSec() > token.expirationTime) {
+            getAccessToken().then((newToken) => {
+                token = newToken;
+                console.log("token expired, got a new one");
+                resolve(token);
+            });
+        } else {
+            resolve(token);
+        }
+    });
+};
+
+type SearchAndPlayBlock = Blocks["searchAndPlay"];
+const searchAndPlay = (Spotify): Block<SearchAndPlayBlock> => ({
+    type: BlockType.Command,
+    arg: { type: ArgumentType.String, defaultValue: "tacos" },
+    text: (searchQuery) => `play music like ${searchQuery}`,
+    operation: async function (searchQuery) {
+        let token = await refreshAccessTokenIfNeeded(Spotify.spotifyToken);
+        Spotify.spotifyToken = token;
+        await Spotify.requestSearch(searchQuery);
+        await Spotify.playTrack();
+    },
+});
+
+type SearchAndPlayWaitBlock = Blocks["searchAndPlayWait"];
+const searchAndPlayWait = (Spotify): Block<SearchAndPlayWaitBlock> => ({
+    type: BlockType.Command,
+    arg: { type: ArgumentType.String, defaultValue: "lauryn hill" },
+    text: (searchQuery) => `play music like ${searchQuery} until done`,
+    operation: async function (searchQuery) {
+        let token = await refreshAccessTokenIfNeeded(Spotify.spotifyToken);
+        Spotify.spotifyToken = token;
+        await Spotify.requestSearch(searchQuery);
+        await Spotify.playTrack();
+        
+        return new Promise<void> (resolve => {
+            setTimeout(() => {
+                resolve();
+            }, Spotify.currentTrackDuration*1000);
+        });
+    },
+});
+
+type StopMusicBlock = Blocks["stopMusic"];
+const stopMusic = (Spotify): Block<StopMusicBlock> => ({
+    type: BlockType.Command,
+    text: `stop the music`,
+    operation: function () {
+        Spotify.player.stop();
+        Spotify.clearTimeouts();
+    },
+});
+
+type TrackData = Blocks["trackData"];
+const getTrackData = (): Block<TrackData> => ({
+    type: BlockType.Reporter,
+    arg: {
+        type: ArgumentType.String,
+        defaultValue: "full",
+        options: ["track", "artist", "album", "full"],
+    },
+    text: (requestedData) => `${requestedData} name`,
+    operation: function (requestedData) {
+        switch (requestedData) {
```

Comment:
Could consider [destructuring](https://www.javascripttutorial.net/es6/javascript-object-destructuring/)  the properties from the extension:

```ts
const { currentTrackName, currentArtistName, currentAlbumName } = this;
```